### ISSUE TITLE
unnested_or_patterns removes parens: top-level or-patterns are not allowed in `let` bindings

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9952]
### Summary

.

### Lint Name

unnested_or_patterns

### Reproducer

I tried this code:

```rust
fn main() {
    let (0 | (1 | _)) = 0;
}
```

I saw this happen:

```

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: top-level or-patterns are not allowed in `let` bindings
 --> src/main.rs:2:9
  |
2 |     let 0 | 1 | _ = 0;
  |         ^^^^^^^^^ help: wrap the pattern in parentheses: `(0 | 1 | _)`

error: aborting due to previous error

Original diagnostics will follow.

warning: unnecessary parentheses around pattern
 --> src/main.rs:2:14
  |
2 |     let (0 | (1 | _)) = 0;
  |              ^     ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
2 -     let (0 | (1 | _)) = 0;
2 +     let (0 | 1 | _) = 0;
  |

warning: unnested or-patterns
 --> src/main.rs:2:9
  |
2 |     let (0 | (1 | _)) = 0;
  |         ^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnested_or_patterns
  = note: requested on the command line with `-W clippy::unnested-or-patterns`
help: nest the patterns
  |
2 |     let 0 | 1 | _ = 0;
  |         ~~~~~~~~~


```

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
